Feature/std23 repeat

joewpeterson opened the pull request.

whaeck committed:
Adding start of repeat_view

joewpeterson committed:
Update repeat view with fully integrated options based on llvm code

Also, include tests for int, float, and string view types

joewpeterson committed:
Alter some spacing that was incorrect

whaeck committed:
Updating ...

joewpeterson committed:
Alter llvm specific unsigned conversion for size to more generic

conversion

whaeck commented on src/tools/std23/views/repeat.hpp:
```diff
@@ -0,0 +1,235 @@
+inline namespace ranges {
+
+template <typename Type>
+inline constexpr bool enable_borrowed_range<std23::ranges::repeat_view<Type>> = forward_range<Type> && enable_borrowed_range<Type>;
```
I don't think this is needed in the case of repeat. For a range adapter like drop which modifies an existing range, it is required though.

joewpeterson replied:
Done. No issues seen when removing this

whaeck commented on src/tools/std23/views/test/repeat.test.cpp:
```diff
@@ -0,0 +1,107 @@
+				CHECK(RepeatedInt.size() == boundLimit);
+				for (int i : RepeatedInt){
+					CHECK(i == repeated);
+				}
```
I would like to see tests for the concepts that the view needs to satisfy:

```
        CHECK( std20::ranges::viewable_range< Range > );

        CHECK( std20::ranges::range< Range > );
        CHECK( std20::ranges::view< Range > );
        CHECK( std20::ranges::sized_range< Range > );
        CHECK( std20::ranges::forward_range< Range > );
        CHECK( std20::ranges::bidirectional_range< Range > );
        CHECK( std20::ranges::random_access_range< Range > );
        CHECK( ! std20::ranges::contiguous_range< Range > );
        CHECK( std20::ranges::common_range< Range > );
```
and tests for the remainder of the view interface. like this:
```
        CHECK( 5 == chunk.size() );

        CHECK( false == chunk.empty() );
        CHECK( true == bool( chunk ) );

        CHECK( std20::ranges::equal( equal, chunk ) );

        CHECK( equal[0] == chunk.front() );
        CHECK( equal[4] == chunk.back() );

        CHECK( 1 == chunk[0] );
        CHECK( 3 == chunk[1] );
        CHECK( 5 == chunk[2] );
        CHECK( 7 == chunk[3] );
        CHECK( 9 == chunk[4] );
```
All other views have tests like these to ensure the views function appropriately.

joewpeterson replied:
Random access issue is fixed with inclusion of >= and <= operators. Pushing this shortly. 

Size Checks are already being done with bound limits with lines like
CHECK(RepeatedInt.size() == boundLimit); 
and with unbound limits with
CHECK( std20::same_as<std::remove_cv_t<decltype(std20::unreachable_sentinel)>, decltype(RepeatedInt.end())>);

Adding to use case test an unbound example as well. Pushing shortly:

whaeck committed:
adding use case

whaeck commented:
I added a use case in the use case test and it seems repeat is not a random access range. I believe it should be.

joewpeterson committed:
Fixing issue with repeat random access

Adding >= and <= operators to make repeat random access. Adding unbound
float type test to use case test. Removing enabled borrow range

whaeck reviewed (approved):
The random access issue is fixed. I'll do a follow up to clean up and add the missing concept tests and view_interface function tests, as well as moving the zip and zip_transform out of production code. After that, we can look at releasing a new version and update ENDFtk.
